### PR TITLE
redirect output when daemonizing and close std* first

### PR DESCRIPTION
**Pull request by Whoops**
### What does this PR do?
This PR redirects input/output to `/dev/null` when daemonizing. It also closes stdin, stdout, and stderr before redirecting them.

### What issues does this PR fix or reference?
This addresses #37934

### Previous Behavior
Salt did not redirect or close the standard inputs/outputs when daemonizing

### Tests written?
No

I'm not sure this is the correct way to go about this. The current behaviour was implemented in [28847a8e](https://github.com/saltstack/salt/commit/28847a8e60d437e80ba64abf2881331aec1e3074) and both the redirect functionality, and daemonizing without it where implemented at the same time, so it was clearly a conscious choice by @thatch45 not to do any redirection.

It appears that `daemonize` is never called anywhere except `deamonize_if` so it's possible that the redirect functionality has never been used.

Also I'm closing the 3 file descriptors before reopening them, because otherwise I was still seeing the behaviour described in #37934, though I'm not sure why that was the case.

I'd welcome any guidance if there are better ideas or preferred approaches.

### CODE REVIEW

**Comment by thatch45** [comment]
Sorry that commit did not have a better explanation. If I recall the issue happens on the second daemonization. so the initial salt minion process daemonizes, and that all goes well, but when it spawns a new job execution salt makes another process and daemonizes that process as well (for many reasons, but mostly so that that process can kill the parent if needed). On that second daemonize we were having issues with stdin and stdout being redirected.
Long story short, this was a hard bug to find and this could break many things :)

**Comment by Whoops** [comment]
Yeah that looked very much to me like someone who knew what they were doing made a decision, which is why I dug through the history. I don't suppose you know how to recreate the bug you were fixing? The last thing I'd want is to send someone chasing a subtle issue, especially something that's already been fixed once.

Also, it looks like I missed the invocation of [daemonize in parsers](https://github.com/saltstack/salt/blob/c734e4d1726a5b688fb904f270bf611f67f53ef2/salt/utils/parsers.py#L1010), which based on how it's `daemonize_if_required` is used, looks like it's doing the work that's related to my troubles. I think the modification to `daemonize_if` wasn't necessary, but I'll have to test and get a better understanding tomorrow, it's late here.

**Comment by cachedout** [comment]
@Whoops Sounds good. Keep us posted. :]

**Comment by Whoops** [comment]
Just got time to look at this again. Whatever is going on, this PR is clearly not the right approach. In the original version, `daemonize` attempts to redirect with the following code:

```python
os.dup2(dev_null.fileno(), sys.stdin.fileno())
os.dup2(dev_null.fileno(), sys.stdout.fileno())
os.dup2(dev_null.fileno(), sys.stderr.fileno())
```

Which looks like it should work. I ran salt-minion through strace like so:
```bash
strace -f -o salt.trace salt-minion -c etc/salt/ -d
```
And located those calls by looking for 3 `dup2`s together. Fortunately that only occurs once (some noise removed):
```
5421  open("/dev/null", O_RDWR)         = 5
5421  fstat(5, {st_mode=S_IFCHR|0666, st_rdev=makedev(1, 3), ...}) = 0
5420  read(4,  <unfinished ...>
5421  fcntl(5, F_GETFD <unfinished ...>
5421  <... fcntl resumed> )             = 0
5421  fcntl(5, F_SETFD, FD_CLOEXEC)     = 0
5421  dup2(5, 3)                        = 3
5421  dup2(5, 1)                        = 1
5421  dup2(5, 2)                        = 2
5421  close(5)                          = 0
```

So it's duping into the file descriptors 1-3, when stdin/out/err should be 0-2. Specifically, it looks like stdin is now fd 3.

A quick check to make sure it isn't python that's wierd:
```python
>>> import sys
>>> sys.stdin.fileno()
0
```

and sure enough if I hard code those calls to 0-2:
```python
with fopen('/dev/null', 'r+') as dev_null: # reopen as dev_null
            os.dup2(dev_null.fileno(), 0)
            os.dup2(dev_null.fileno(), 1)
            os.dup2(dev_null.fileno(), 2)
```
now the trace looks as expected:
```
open("/dev/null", O_RDWR)         = 5
6289  fstat(5, {st_mode=S_IFCHR|0666, st_rdev=makedev(1, 3), ...}) = 0
6289  fcntl(5, F_GETFD)                 = 0
6289  fcntl(5, F_SETFD, FD_CLOEXEC)     = 0
6289  dup2(5, 0)                        = 0
6289  dup2(5, 1)                        = 1
6289  dup2(5, 2)                        = 2
```

And the problem from #37934 is gone.

So somewhere somehow in minion startup, sys.stdin is reopened and gets fd 3. I'm going to try to track down where and see if it makes sense. Hard coding the fd numbers like above would fix that specific problem, but I'm not sure how/why stdin got reopened, so I'm not sure if there's another underlying problem.

**Comment by Whoops** [comment]
fd 3 (aka the current stdin) is `/dev/null` at that point in time, so something reopened/replaced it. Still trying to figure out when/how.

**Comment by Whoops** [comment]
Ok, it's the multiprocessing module itself that's doing it. Before the subprocess is started [here](https://github.com/saltstack/salt/blob/c734e4d1726a5b688fb904f270bf611f67f53ef2/salt/scripts.py#L190) the stdin fd is 0, however by the time `minion_process` is invoked in the subprocess, it's fd 3. It remains fd 0 in the parent.

Why the multiprocessing module does this, I'm not sure (though I'm curious enough that I might go spelunking to find out). Given that it's coming from the multiprocessing module though, I think the sanest thing is probably just to hard-code those fd numbers. I've updated the pull request to do so.

I tried using `sys.__stdin__.fileno()` however I got `ValueError: I/O operation on closed file`, so it I don't see a better way.

**Comment by Whoops** [comment]
This might be unintended in `multiprocessing` too.  Looking at [_bootstrap](https://hg.python.org/releases/2.7.11/file/tip/Lib/multiprocessing/process.py#l238) stdin is closed and reopened with the following:
```python
sys.stdin.close()
sys.stdin = open(os.devnull)
```
With stdin closed one might expect the next `open` to get fd 0, however sys.stdin.close doesn't appear to actually close anything. A quick test with `strace` and this script:
```python
import sys
import os

sys.stdin.close()
sys.stdin = open(os.devnull)
```

Shows that fd 0 is never closed, and `/dev/null` is opened with fd 3 (`open("/dev/null", O_RDONLY)             = 3`).

However if it's changed to
```python
import sys
import os

os.close(sys.stdin.fileno())
sys.stdin = open(os.devnull)
```
Then fd 0 is indeed closed and reopened:
```
close(0)                                = 0
open("/dev/null", O_RDONLY)             = 0
```

Chasing that rabbit hole is a project for another day. I think based on that, this pull request might be the best approach.

**Comment by cachedout** [comment]
@thatch45 Did you see the above comments?

**Comment by thatch45** [comment]
Lets merge it in!
Sorry I am so late here, it has been a beyond crazy couple of weeks!
